opt: add lock operator

michae2 opened the pull request:
**execbuilder: change error message to match PostgreSQL**

Release note (sql change): Change the following error message:

```
statement error cannot execute FOR UPDATE in a read-only transaction
```

to this to match PostgreSQL:

```
statement error cannot execute SELECT FOR UPDATE in a read-only transaction
```

**opt: add lock operator**

Add a new implementation of `SELECT FOR UPDATE` and `SELECT FOR SHARE`
statements. Instead of locking during the initial row fetch, this new
implementation constructs a `Lock` operator on the top of the query plan
which performs the locking phase using a locking semi-join lookup.

During optbuilder we build plans with both `Lock` operators and
initial-row-fetch locking. During execbuilder we decide which
implementation to use based on the isolation level and whether
`optimizer_use_lock_op_for_serializable` is set. If the new
implementation is chosen, `Lock` operators become locking
semi-LookupJoins.

In some cases these new plans will have superfluous lookup joins. A
future PR will optimize away some of these superfluous lookup joins.

Fixes: #57031, #75457

Epic: CRDB-25322

Release note (sql change): Add a new session variable,
`optimizer_use_lock_op_for_serializable`, which when set enables a new
implementation of `SELECT FOR UPDATE`. This new implementation of
`SELECT FOR UPDATE` acquires row locks *after* any joins and filtering,
and always acquires row locks on the primary index of the table being
locked. This more closely matches `SELECT FOR UPDATE` behavior in
PostgreSQL, but at the cost of more round trips from gateway node to
replica leaseholder.

Under read committed isolation (and other isolation levels weaker than
serializable) we will always use this new implementation of `SELECT FOR
UPDATE` regardless of the value of
`optimizer_use_lock_op_for_serializable` to ensure correctness.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/111662)
<!-- Reviewable:end -->

DrewKimball reviewed (commented):
I left some comments. Feel free to disregard any or all as you keep iterating.

> Reviewed 2 of 2 files at r1, 10 of 10 files at r2, 17 of 17 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/111662#-:-Nfwiemn1DNBY3gqF-Ib:bo6mkg2)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @michae2)

___
*[`pkg/sql/opt/exec/execbuilder/builder.go` line 318 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/111662#-NfwIzz9C68Tiionyqjc:-NfwIzz9C68Tiionyqjd:bpy90l2) ([raw file](https://github.com/cockroachdb/cockroach/blob/478eed5ae0be3d72b3ab3104564b9fff8c558133/pkg/sql/opt/exec/execbuilder/builder.go#L318)):*
> ```Go
> 	b.allowInsertFastPath = b.allowInsertFastPath && canAutoCommit
> 
> 	b.allowLocking = b.evalCtx.TxnIsoLevel == isolation.Serializable &&
> ```

Do we not want to set this for read committed?
___
*[`pkg/sql/opt/exec/execbuilder/mutation.go` line 1134 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/111662#-NfwJTOP5OJtQq3jgYIq:-NfwJTOP5OJtQq3jgYIr:b-5dmkm7) ([raw file](https://github.com/cockroachdb/cockroach/blob/478eed5ae0be3d72b3ab3104564b9fff8c558133/pkg/sql/opt/exec/execbuilder/mutation.go#L1134)):*
> ```Go
> 	if !lock.Locking.IsLocking() ||
> 		!b.evalCtx.Settings.Version.IsActive(b.ctx, clusterversion.V23_2) ||
> 		(b.evalCtx.TxnIsoLevel == isolation.Serializable &&
> ```

[nit] Maybe you could split out the Serializable conditions to make it more clear?
___
*[`pkg/sql/opt/exec/execbuilder/mutation.go` line 1145 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/111662#-NfweAFL2CMblWqfDZv6:-NfweAFL2CMblWqfDZv7:b-ybaazt) ([raw file](https://github.com/cockroachdb/cockroach/blob/478eed5ae0be3d72b3ab3104564b9fff8c558133/pkg/sql/opt/exec/execbuilder/mutation.go#L1145)):*
> ```Go
> 		Input: lock.Input,
> 		LookupJoinPrivate: memo.LookupJoinPrivate{
> 			JoinType:              opt.InnerJoinOp,
> ```

Could this be a `SemiJoin`? It looked like we don't project columns from the lookups, and we're joining on a unique key anyway...
___
*[`pkg/sql/opt/exec/execbuilder/mutation.go` line 1158 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/111662#-Nfwevad4zGzXhqMgqpb:-Nfwevad4zGzXhqMgqpc:bg7pv1l) ([raw file](https://github.com/cockroachdb/cockroach/blob/478eed5ae0be3d72b3ab3104564b9fff8c558133/pkg/sql/opt/exec/execbuilder/mutation.go#L1158)):*
> ```Go
> 	if !b.allowLocking {
> 		b.allowLocking = true
> 		defer func() { b.allowLocking = false }()
> ```

Can't lock operators be nested?
___
*[`pkg/sql/opt/ops/README.md` line 76 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/111662#-NfwNBf42C90WC_h42Tu:-NfwNBf42C90WC_h42Tv:b-twqble) ([raw file](https://github.com/cockroachdb/cockroach/blob/4494e945a9070862cfae1b668b3f8b667c41be17/pkg/sql/opt/ops/README.md#L76)):*
> ```Markdown
>    expression in the first child. Such expressions must implement a
>    `WithBindingID()` method.
> ```

Was this intentional?
___
*[`pkg/sql/opt/optbuilder/join.go` line 36 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/111662#-Nfwe-TN5TeOh_Ecpidu:-Nfwe-TN5TeOh_Ecpidv:bxk7eco) ([raw file](https://github.com/cockroachdb/cockroach/blob/478eed5ae0be3d72b3ab3104564b9fff8c558133/pkg/sql/opt/optbuilder/join.go#L36)):*
> ```Go
> 	join *tree.JoinTableExpr, lockCtx lockingContext, inScope *scope,
> ) (outScope *scope) {
> 	// TODO(michae2): poison the lockCtx for the null-extended side(s) if this is
> ```

The tracking issue is https://github.com/cockroachdb/cockroach/issues/97434
___
*[`pkg/sql/opt/optbuilder/locking.go` line 94 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/111662#-NfwTZOfBJyJzA9QNSNC:-NfwTZOfBJyJzA9QNSND:bdtoyof) ([raw file](https://github.com/cockroachdb/cockroach/blob/478eed5ae0be3d72b3ab3104564b9fff8c558133/pkg/sql/opt/optbuilder/locking.go#L94)):*
> ```Go
> 			panic(errors.AssertionFailedf("fetchScope did not have column %d", ord))
> 		}
> 		scopeCol := fetchScope.cols[ord]
> ```

Retrieving the column like this relies on the scope having the columns in the same order as the locked table. Maybe that's true in practice, but it seems fragile. You can get the column ID directly from the `TableMeta/TableID` - see an example here: https://github.com/cockroachdb/cockroach/blob/a267928ac466361dbd624ae8a2e277a9aab59f80/pkg/sql/opt/xform/join_funcs.go#L1024-L1033.
___
*[`pkg/sql/opt/optbuilder/locking.go` line 272 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/111662#-NfwhlWNBblHc6622Ke6:-NfwhlWNBblHc6622Ke7:bc96emg) ([raw file](https://github.com/cockroachdb/cockroach/blob/478eed5ae0be3d72b3ab3104564b9fff8c558133/pkg/sql/opt/optbuilder/locking.go#L272)):*
> ```Go
> 
> 	for i := range lockScope.cols {
> 		b.addOrderByOrDistinctOnColumn(inScope, projectionsScope, lockScope, &lockScope.cols[i])
> ```

I don't think this is necessary, since the locking columns are only ever references to table columns, rather than expressions that need to be built. You could probably combine this step with `analyzeLockArgs` and just skip the step with `lockScope`, but I think it's fine either way.
___
*[`pkg/sql/opt/optbuilder/locking.go` line 366 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/111662#-NfwZZlX5a-w5z-7wHHv:-NfwZZlX5a-w5z-7wHHw:b-8j1iwb) ([raw file](https://github.com/cockroachdb/cockroach/blob/478eed5ae0be3d72b3ab3104564b9fff8c558133/pkg/sql/opt/optbuilder/locking.go#L366)):*
> ```Go
> 	}
> 	//fmt.Println("buildLock, keyCols:", lb.keyCols, "scope.cols", inScope.cols, "scope.extraCols", inScope.extraCols)
> 	for _, col := range inScope.cols {
> ```

You can use `colSetWithExtraCols` here :)
___
*[`pkg/sql/opt/optbuilder/locking.go` line 381 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/111662#-NfwZGAy-u38GZKqi_kn:-NfwZGAy-u38GZKqi_ko:b-pe0vqd) ([raw file](https://github.com/cockroachdb/cockroach/blob/478eed5ae0be3d72b3ab3104564b9fff8c558133/pkg/sql/opt/optbuilder/locking.go#L381)):*
> ```Go
> 		}
> 	}
> 	inScope.expr = b.factory.ConstructLock(inScope.expr, private)
> ```

Did you find a reason why not to directly construct a LookupJoinExpr here? It seems like doing that could simplify the logic down the line. Especially since we plan to make locking an enforced property in the future, with `LockExpr` only requiring the property.
___
*[`pkg/sql/opt/optbuilder/select.go` line 160 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/111662#-NfwiUyc6e_1Z7QmhMM7:-NfwiUyc6e_1Z7QmhMM8:bulkrsi) ([raw file](https://github.com/cockroachdb/cockroach/blob/478eed5ae0be3d72b3ab3104564b9fff8c558133/pkg/sql/opt/optbuilder/select.go#L160)):*
> ```Go
> 			)
> 
> 			if lockCtx.locking.isSet() {
> ```

Should this just be inlined into `buildScan`?


<!-- Sent from Reviewable.io -->

michae2 committed:
execbuilder: change error message to match PostgreSQL

Release note (sql change): Change the following error message:

```
statement error cannot execute FOR UPDATE in a read-only transaction
```

to this to match PostgreSQL:

```
statement error cannot execute SELECT FOR UPDATE in a read-only transaction
```

michae2 committed:
opt: add lock operator

Add a new implementation of `SELECT FOR UPDATE` and `SELECT FOR SHARE`
statements. Instead of locking during the initial row fetch, this new
implementation constructs a `Lock` operator on the top of the query plan
which performs the locking phase using a locking semi-join lookup.

During optbuilder we build plans with both `Lock` operators and
initial-row-fetch locking. During execbuilder we decide which
implementation to use based on the isolation level and whether
`optimizer_use_lock_op_for_serializable` is set. If the new
implementation is chosen, `Lock` operators become locking
semi-LookupJoins.

In some cases these new plans will have superfluous lookup joins. A
future PR will optimize away some of these superfluous lookup joins.

Fixes: #57031, #75457

Epic: CRDB-25322

Release note (sql change): Add a new session variable,
`optimizer_use_lock_op_for_serializable`, which when set enables a new
implementation of `SELECT FOR UPDATE`. This new implementation of
`SELECT FOR UPDATE` acquires row locks *after* any joins and filtering,
and always acquires row locks on the primary index of the table being
locked. This more closely matches `SELECT FOR UPDATE` behavior in
PostgreSQL, but at the cost of more round trips from gateway node to
replica leaseholder.

Under read committed isolation (and other isolation levels weaker than
serializable) we will always use this new implementation of `SELECT FOR
UPDATE` regardless of the value of
`optimizer_use_lock_op_for_serializable` to ensure correctness.

michae2 reviewed (commented):
Thanks for the comments! I'm still ironing out a few small issues but this is RFAL.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/111662#-:-NgLlClaCv_X63PmyqqQ:b6r1t6q)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball)

___
*[`pkg/sql/opt/exec/execbuilder/builder.go` line 318 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/111662#-NfwIzz9C68Tiionyqjc:-NgLeeT853sfn0Hhqi_d:b4fmf0w) ([raw file](https://github.com/cockroachdb/cockroach/blob/478eed5ae0be3d72b3ab3104564b9fff8c558133/pkg/sql/opt/exec/execbuilder/builder.go#L318)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Do we not want to set this for read committed?
</blockquote></details>

I've reversed the logic here, so now we're setting this (newly-named) boolean for read committed.
___
*[`pkg/sql/opt/exec/execbuilder/mutation.go` line 1145 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/111662#-NfweAFL2CMblWqfDZv6:-NgLeyWEACABheJ2dchO:b-hlk7bf) ([raw file](https://github.com/cockroachdb/cockroach/blob/478eed5ae0be3d72b3ab3104564b9fff8c558133/pkg/sql/opt/exec/execbuilder/mutation.go#L1145)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Could this be a `SemiJoin`? It looked like we don't project columns from the lookups, and we're joining on a unique key anyway...
</blockquote></details>

Ah, good catch! Done.
___
*[`pkg/sql/opt/exec/execbuilder/mutation.go` line 1158 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/111662#-Nfwevad4zGzXhqMgqpb:-NgLf0Ce1f9cg7mhBYG_:b-df2yy5) ([raw file](https://github.com/cockroachdb/cockroach/blob/478eed5ae0be3d72b3ab3104564b9fff8c558133/pkg/sql/opt/exec/execbuilder/mutation.go#L1158)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Can't lock operators be nested?
</blockquote></details>

Yes, this was broken. Good catch. Now we're marking these `opt.Locking` structs by setting `Forced` when they come from here, and keeping `onlyForcedLocking` the same.
___
*[`pkg/sql/opt/ops/README.md` line 76 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/111662#-NfwNBf42C90WC_h42Tu:-NgLfXTG69Gmt5YIcLNV:btv2dmy) ([raw file](https://github.com/cockroachdb/cockroach/blob/4494e945a9070862cfae1b668b3f8b667c41be17/pkg/sql/opt/ops/README.md#L76)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Was this intentional?
</blockquote></details>

Yes (emacs highlights extra trailing newlines, so this was annoying me).
___
*[`pkg/sql/opt/optbuilder/join.go` line 36 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/111662#-Nfwe-TN5TeOh_Ecpidu:-NgLgH2xAEo5DPlX8UNH:b-7rl0jd) ([raw file](https://github.com/cockroachdb/cockroach/blob/478eed5ae0be3d72b3ab3104564b9fff8c558133/pkg/sql/opt/optbuilder/join.go#L36)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

The tracking issue is https://github.com/cockroachdb/cockroach/issues/97434
</blockquote></details>

done.
___
*[`pkg/sql/opt/optbuilder/locking.go` line 94 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/111662#-NfwTZOfBJyJzA9QNSNC:-NgLgNKJ3MAtK-Dru0jW:bz6u692) ([raw file](https://github.com/cockroachdb/cockroach/blob/478eed5ae0be3d72b3ab3104564b9fff8c558133/pkg/sql/opt/optbuilder/locking.go#L94)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Retrieving the column like this relies on the scope having the columns in the same order as the locked table. Maybe that's true in practice, but it seems fragile. You can get the column ID directly from the `TableMeta/TableID` - see an example here: https://github.com/cockroachdb/cockroach/blob/a267928ac466361dbd624ae8a2e277a9aab59f80/pkg/sql/opt/xform/join_funcs.go#L1024-L1033.
</blockquote></details>

Great point, thank you for the example. Changed to get the column ID from the TableMeta.
___
*[`pkg/sql/opt/optbuilder/locking.go` line 272 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/111662#-NfwhlWNBblHc6622Ke6:-NgLga2hCDAJpSjcIdR1:b-ya9n2z) ([raw file](https://github.com/cockroachdb/cockroach/blob/478eed5ae0be3d72b3ab3104564b9fff8c558133/pkg/sql/opt/optbuilder/locking.go#L272)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

I don't think this is necessary, since the locking columns are only ever references to table columns, rather than expressions that need to be built. You could probably combine this step with `analyzeLockArgs` and just skip the step with `lockScope`, but I think it's fine either way.
</blockquote></details>

I was wondering this, thank you. I don't think we can combine with `analyzeLockArgs` because that needs to run after `analyze
___
*[`pkg/sql/opt/optbuilder/locking.go` line 366 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/111662#-NfwZZlX5a-w5z-7wHHv:-NgLkXFE0FyxjWgt94zj:b-hektvl) ([raw file](https://github.com/cockroachdb/cockroach/blob/478eed5ae0be3d72b3ab3104564b9fff8c558133/pkg/sql/opt/optbuilder/locking.go#L366)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

You can use `colSetWithExtraCols` here :)
</blockquote></details>

Gah, I am such a newb. :slightly_smiling_face: Done.

Actually, there might be a small bug here. It should work with `colSet` instead of `colSetWithExtraCols`. Stay tuned...
___
*[`pkg/sql/opt/optbuilder/locking.go` line 381 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/111662#-NfwZGAy-u38GZKqi_kn:-NgLhyj41-jvqrU1ERCV:b-7gf293) ([raw file](https://github.com/cockroachdb/cockroach/blob/478eed5ae0be3d72b3ab3104564b9fff8c558133/pkg/sql/opt/optbuilder/locking.go#L381)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Did you find a reason why not to directly construct a LookupJoinExpr here? It seems like doing that could simplify the logic down the line. Especially since we plan to make locking an enforced property in the future, with `LockExpr` only requiring the property.
</blockquote></details>

Two reasons:
1. Because we want `LockExpr` to act as an optimization barrier during exploration, and I don't think `LookupJoinExpr` would.
2. In 23.2 I'm trying to only use `LockExpr` under read committed isolation to reduce risk. This means that in optbuilder we don't know for certain which locking implementation we will use (because we could prepare a `SELECT FOR UPDATE` statement under serializable isolation and then execute it under read committed isolation or vice versa). So we need to wait until execbuilder, when we know for certain, to either convert the `LockExpr` to `LookupJoinExpr` or elide it.

This will become simpler when we fully commit to LockExpr and make locking an enforced property. Then execbuilder will go back to simply eliding the `LockExpr`.
___
*[`pkg/sql/opt/optbuilder/select.go` line 160 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/111662#-NfwiUyc6e_1Z7QmhMM7:-NgLgxsQ53PIcUDVjf6U:b-c70bw6) ([raw file](https://github.com/cockroachdb/cockroach/blob/478eed5ae0be3d72b3ab3104564b9fff8c558133/pkg/sql/opt/optbuilder/select.go#L160)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Should this just be inlined into `buildScan`?
</blockquote></details>

I considered it, but out of over a dozen calls to `buildScan` only these two calls deal with locking contexts.


<!-- Sent from Reviewable.io -->

msirek reviewed (commented):
> Reviewed 6 of 23 files at r4, 5 of 40 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/111662)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball, @mgartner, and @michae2)

___
*[`pkg/sql/opt/exec/execbuilder/mutation.go` line 1145 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/111662#-NfweAFL2CMblWqfDZv6:-NgM4ftjDUrYI4hJohjf:bjvmwr) ([raw file](https://github.com/cockroachdb/cockroach/blob/478eed5ae0be3d72b3ab3104564b9fff8c558133/pkg/sql/opt/exec/execbuilder/mutation.go#L1145)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Ah, good catch! Done.
</blockquote></details>

Just wondering why the join is built in the execbuilder instead of the optbuilder. Is it because you want to make sure the lookup join is never removed by some rule firing?  I can imagine it might mess up the stats if we don't recognize the lookup join is essentially an index join.  I'm just wondering if placing more logic in the execbuilder will make it more difficult to maintain in the long run, instead of being a straightforward translation of what's in the optimized expression tree.


<!-- Sent from Reviewable.io -->

msirek reviewed (commented):
Great work!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/111662#-:-NgM7uH-09BbjD7kkqMW:b-9dwnzc)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball, @mgartner, and @michae2)



<!-- Sent from Reviewable.io -->

DrewKimball reviewed (commented):
This is looking pretty good. My main concern right now is plan regressions for serializable isolation - there are a couple places where we're stopping column-pruning or join-elimination rules from firing when they did before, and I don't think it's necessary:

>In 23.2 I'm trying to only use `LockExpr` under read committed isolation to reduce risk. This means that in optbuilder we don't know for certain which locking implementation we will use (because we could prepare a `SELECT FOR UPDATE` statement under serializable isolation and then execute it under read committed isolation or vice versa). So we need to wait until execbuilder, when we know for certain, to either convert the `LockExpr` to `LookupJoinExpr` or elide it.

We should be invalidating the memo and re-planning when the isolation level changes for prepared statements and regular ones. So, it should be safe to expect the same isolation level during execution as we observe during planning, even for prepared statements. So we should be able to avoid building the lock operator under the default settings. Maybe that will simplify some of your reasoning too?

> Reviewed 23 of 23 files at r4, 40 of 40 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/111662#-:-NgMALZD6fB1OE7S-zDU:b8dtvaj)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner and @michae2)

___
*[`pkg/sql/opt/locking.go` line 74 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/111662#-NgM11DtDpJTpnWSONnC:-NgM11DtDpJTpnWSONnD:bw90fd7) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ec4ae291b0f316f43449610f2eb5558641d5c9d/pkg/sql/opt/locking.go#L74)):*
> ```Go
> 	// locking is disabled under weaker isolation levels. It is used to
> 	// distinguish Lock-operator locking from initial-row-fetch locking.
> 	Forced bool
> ```

[nit] `Forced` was a bit confusing. How would you feel about `UseLockOp` and `ForceUseLockOperator`, or something like that? Or maybe `UseExplicitLocking`
___
*[`pkg/sql/opt/exec/execbuilder/mutation.go` line 1180 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/111662#-NgLuLrA-wlrfxxwhtlu:-NgLuLrA-wlrfxxwhtlv:b-mky8ec) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ec4ae291b0f316f43449610f2eb5558641d5c9d/pkg/sql/opt/exec/execbuilder/mutation.go#L1180)):*
> ```Go
> 		},
> 	}
> 	memo.CopyLockGroupIntoLookupJoin(lock, join)
> ```

Probably what you want here is `AddLookupJoinToGroup`.
___
*[`pkg/sql/opt/exec/execbuilder/testdata/select_for_update_read_committed` line 131 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/111662#-NgM81rB6jq9PhRgiVTV:-NgM81rB6jq9PhRgiVTW:b-nwco8m) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ec4ae291b0f316f43449610f2eb5558641d5c9d/pkg/sql/opt/exec/execbuilder/testdata/select_for_update_read_committed#L131)):*
> ```
>                   table: supermarket@supermarket_pkey
>                   spans: /"matilda"/0
>                   locking strength: for update
> ```

Why is the initial-scan locking still printed here? I notice it's been removed in some places, but not others.
___
*[`pkg/sql/opt/xform/testdata/rules/limit` line 907 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/111662#-NgM9kv8BEQzsZefykdB:-NgM9kv8BEQzsZefykdC:b-dlahar) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ec4ae291b0f316f43449610f2eb5558641d5c9d/pkg/sql/opt/xform/testdata/rules/limit#L907)):*
> ```
> SELECT * FROM kuv ORDER BY u LIMIT 5 FOR UPDATE
> ----
> sort
> ```

We should add the lock operator to the list in `opt/ordering/ordering/go`, since it should be able to pass through its input ordering. See `opt/ordering/lookup_join.go`. Maybe that doesn't have to happen in this PR, though.


<!-- Sent from Reviewable.io -->

michae2 reviewed (commented):
Thanks for pointing this out, it did simplify things! The "forced" locking stuff is gone, as are the plan regressions for serializable.

I'm still fighting with the optimizer about a few things, but this is closer now. Also, needs more tests.

First commit is #112082 which can be ignored.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/111662#-:-NgNVZ8k9jPoSM7YH9-E:b-oco2s7)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball, @mgartner, and @msirek)

___
*[`pkg/sql/opt/exec/execbuilder/mutation.go` line 1134 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/111662#-NfwJTOP5OJtQq3jgYIq:-NgNWI8BDr3N5tJ80bf5:b6xb7ms) ([raw file](https://github.com/cockroachdb/cockroach/blob/478eed5ae0be3d72b3ab3104564b9fff8c558133/pkg/sql/opt/exec/execbuilder/mutation.go#L1134)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

[nit] Maybe you could split out the Serializable conditions to make it more clear?
</blockquote></details>

I changed this to simply call `buildLocking` (which is what it was emulating).
___
*[`pkg/sql/opt/exec/execbuilder/mutation.go` line 1145 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/111662#-NfweAFL2CMblWqfDZv6:-NgNWTIK6rZvlghKcpm5:b-ki1581) ([raw file](https://github.com/cockroachdb/cockroach/blob/478eed5ae0be3d72b3ab3104564b9fff8c558133/pkg/sql/opt/exec/execbuilder/mutation.go#L1145)):*
<details><summary><i>Previously, msirek (Mark Sirek) wrote…</i></summary><blockquote>

Just wondering why the join is built in the execbuilder instead of the optbuilder. Is it because you want to make sure the lookup join is never removed by some rule firing?  I can imagine it might mess up the stats if we don't recognize the lookup join is essentially an index join.  I'm just wondering if placing more logic in the execbuilder will make it more difficult to maintain in the long run, instead of being a straightforward translation of what's in the optimized expression tree.
</blockquote></details>

Yes, the reason is to avoid matching any rules that would push something above or below the lookup join, or remove the lookup join.

I agree that it's a little funny to create the lookup join in execbuilder. Maybe one way to justify it is to consider lookup join as merely the implementation of `Lock`, and then the call to `buildLookupJoin` is merely sharing code, and not some sneaky attempt to modify the plan in exec builder. Similar to how you were calling `ConstructOrdinality` from `buildUniqueKey` in your [prototype](https://github.com/cockroachdb/cockroach/pull/109903/files#diff-366e1ec6fe597a3ff84d46aa04c4a2753f1b0911d1d6dac6868096280480cbbe). But maybe I'm just arguing semantics.

The long-term plan is to make locking a physical property, make `Lock` require the physical property, and make lookup join an enforcer for the physical property. I think in that plan we still build `Lock` during optbuilder, and it still acts as an optimization barrier during exploration. But I didn't quite get there in this PR.
___
*[`pkg/sql/opt/exec/execbuilder/mutation.go` line 1180 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/111662#-NgLuLrA-wlrfxxwhtlu:-NgNZjvIBJwAq73UVi75:b-4o8qar) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ec4ae291b0f316f43449610f2eb5558641d5c9d/pkg/sql/opt/exec/execbuilder/mutation.go#L1180)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Probably what you want here is `AddLookupJoinToGroup`.
</blockquote></details>

I've been thinking about this, but I don't actually want to add the lookup join to the memo. Just want to create it to simplify calling `buildLookupJoin`. Maybe I'm being quixotic.
___
*[`pkg/sql/opt/exec/execbuilder/testdata/select_for_update_read_committed` line 131 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/111662#-NgM81rB6jq9PhRgiVTV:-NgN_3lg5-9Pq19HGR-D:bmph1o) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ec4ae291b0f316f43449610f2eb5558641d5c9d/pkg/sql/opt/exec/execbuilder/testdata/select_for_update_read_committed#L131)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Why is the initial-scan locking still printed here? I notice it's been removed in some places, but not others.
</blockquote></details>

Good eye! This is because `forceForUpdateLocking` is [set](https://github.com/cockroachdb/cockroach/blob/7e1f909e9f2ddcef2e418d6c2df11f3139a2b85a/pkg/sql/opt/exec/execbuilder/mutation.go#L38-L42) for everything below the update. I might have to turn off implicit update locking under read committed for this reason, but I'd rather do it in a follow-up PR.
___
*[`pkg/sql/opt/xform/testdata/rules/limit` line 907 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/111662#-NgM9kv8BEQzsZefykdB:-NgNa5JIE3V0ReOev92b:bxaj6jr) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ec4ae291b0f316f43449610f2eb5558641d5c9d/pkg/sql/opt/xform/testdata/rules/limit#L907)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

We should add the lock operator to the list in `opt/ordering/ordering/go`, since it should be able to pass through its input ordering. See `opt/ordering/lookup_join.go`. Maybe that doesn't have to happen in this PR, though.
</blockquote></details>

Done. (Hopefully I did it correctly... I think it's pretty simple, since we just pass through the ordering.)
___
*[`pkg/sql/opt/locking.go` line 74 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/111662#-NgM11DtDpJTpnWSONnC:-NgNaIUTDWBN-frvMXZj:bkal96b) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ec4ae291b0f316f43449610f2eb5558641d5c9d/pkg/sql/opt/locking.go#L74)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

[nit] `Forced` was a bit confusing. How would you feel about `UseLockOp` and `ForceUseLockOperator`, or something like that? Or maybe `UseExplicitLocking`
</blockquote></details>

Removed it (thanks to your suggestion!)


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (commented):
>The "forced" locking stuff is gone, as are the plan regressions for serializable.

Nice! Glad you managed to fix those regressions. I think there might(?) be a bug with propagating key columns for nested locks, but other than that I'm about ready to stamp. BTW, are there logic tests for locking under read committed somewhere?

> Reviewed 38 of 38 files at r6, 3 of 3 files at r7, 6 of 27 files at r8.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/111662#-:-NgNfTdT0awsuZ1m0lJ5:biiky4i)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner and @michae2)

___
*[`pkg/sql/opt/exec/execbuilder/mutation.go` line 1180 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/111662#-NgLuLrA-wlrfxxwhtlu:-NgNdCoa7KRsc--jFtDy:b-njhc4g) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ec4ae291b0f316f43449610f2eb5558641d5c9d/pkg/sql/opt/exec/execbuilder/mutation.go#L1180)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

I've been thinking about this, but I don't actually want to add the lookup join to the memo. Just want to create it to simplify calling `buildLookupJoin`. Maybe I'm being quixotic.
</blockquote></details>

Well, it's hacky, but it seems correct. But maybe we could use `MemoizeLookupJoin` instead - that won't add it to the query plan tree, and there's precedent for using it for throwaway calculation: https://github.com/cockroachdb/cockroach/blob/ce1d751513af407890a72aa4d88053feb31fb776/pkg/sql/opt/memo/statistics_builder.go#L2872-L2874

It would be less efficient, but I don't think it's worth worrying about since it's only once per lock operator. I'd be alright with keeping the current code for now, though, up to you.
___
*[`pkg/sql/opt/memo/logical_props_builder.go` line 1530 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/111662#-NgNZW75DAkobzOuQ74f:-NgNZW75DAkobzOuQ74g:b9axheu) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef846d1a2957fe175d6a005edc62ec29df565a2f/pkg/sql/opt/memo/logical_props_builder.go#L1530)):*
> ```Go
> 	rel.NotNullCols.IntersectionWith(lock.Cols)
> 	rel.FuncDeps.CopyFrom(&inputProps.FuncDeps)
> 	rel.FuncDeps.ProjectCols(rel.OutputCols)
> ```

As I mentioned elsewhere, I think we should simplify things as much as possible and have the operator just logically pass through columns (apart from possible filtering for skip locked). These lines wouldn't be incorrect in that case, just unnecessary.
___
*[`pkg/sql/opt/optbuilder/locking.go` line 366 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/111662#-NfwZZlX5a-w5z-7wHHv:-NgNYIzaDMSmuXF_8iOX:b-tykprj) ([raw file](https://github.com/cockroachdb/cockroach/blob/478eed5ae0be3d72b3ab3104564b9fff8c558133/pkg/sql/opt/optbuilder/locking.go#L366)):*
>`colSet` instead of `colSetWithExtraCols`

Aren't primary key cols for lock operators potentially propagated as extra cols? What happens if lock operators are nested? I think it might be best to simplify things by just always returning exactly the set of input columns. I think that will "just work" - a projection will be added on top if necessary to remove extra columns after the lock. It won't be totally efficient, but that's probably ok.

If I missed something and the nested-locks case isn't a bug, the above request is just a nit that we can worry about later.
___
*[`pkg/sql/opt/ordering/mutation.go` line 77 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/111662#-NgNahDM89bD4ZjDFAs_:-NgNahDNAScSROZ_6M2u:b-y3hukm) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef846d1a2957fe175d6a005edc62ec29df565a2f/pkg/sql/opt/ordering/mutation.go#L77)):*
>Need to remember the ordering to set in lookupjoin?

This should just work :)
___
*[`pkg/sql/opt/xform/testdata/rules/limit` line 907 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/111662#-NgM9kv8BEQzsZefykdB:-NgNbH6r6MAWY1dvv5wU:b48fw2o) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ec4ae291b0f316f43449610f2eb5558641d5c9d/pkg/sql/opt/xform/testdata/rules/limit#L907)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Done. (Hopefully I did it correctly... I think it's pretty simple, since we just pass through the ordering.)
</blockquote></details>

The ordering code looks good to me!


<!-- Sent from Reviewable.io -->

michae2 reviewed (commented):
> BTW, are there logic tests for locking under read committed somewhere?

There are some in [pkg/sql/logictest/testdata/logic_test/read_committed](https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/logictest/testdata/logic_test/read_committed) as well as in [pkg/sql/opt/exec/execbuilder/testdata/select_for_update_read_committed](https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/opt/exec/execbuilder/testdata/select_for_update_read_committed) but I'm working on adding more.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/111662#-:-NgNmKyB6moxtax6RZ2U:buqcvs5)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball and @mgartner)

___
*[`pkg/sql/opt/exec/execbuilder/mutation.go` line 1180 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/111662#-NgLuLrA-wlrfxxwhtlu:-NgNn9eu--Cg7JHAt92E:b33wmln) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ec4ae291b0f316f43449610f2eb5558641d5c9d/pkg/sql/opt/exec/execbuilder/mutation.go#L1180)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Well, it's hacky, but it seems correct. But maybe we could use `MemoizeLookupJoin` instead - that won't add it to the query plan tree, and there's precedent for using it for throwaway calculation: https://github.com/cockroachdb/cockroach/blob/ce1d751513af407890a72aa4d88053feb31fb776/pkg/sql/opt/memo/statistics_builder.go#L2872-L2874

It would be less efficient, but I don't think it's worth worrying about since it's only once per lock operator. I'd be alright with keeping the current code for now, though, up to you.
</blockquote></details>

Oh, nice. Done.
___
*[`pkg/sql/opt/memo/logical_props_builder.go` line 1530 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/111662#-NgNZW75DAkobzOuQ74f:-NgNnLbtDdQIOtKqqvWp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef846d1a2957fe175d6a005edc62ec29df565a2f/pkg/sql/opt/memo/logical_props_builder.go#L1530)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

As I mentioned elsewhere, I think we should simplify things as much as possible and have the operator just logically pass through columns (apart from possible filtering for skip locked). These lines wouldn't be incorrect in that case, just unnecessary.
</blockquote></details>

Done.
___
*[`pkg/sql/opt/optbuilder/locking.go` line 366 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/111662#-NfwZZlX5a-w5z-7wHHv:-NgNnQuW7yAVZjuu_o4m:b-7ez6mr) ([raw file](https://github.com/cockroachdb/cockroach/blob/478eed5ae0be3d72b3ab3104564b9fff8c558133/pkg/sql/opt/optbuilder/locking.go#L366)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

>`colSet` instead of `colSetWithExtraCols`

Aren't primary key cols for lock operators potentially propagated as extra cols? What happens if lock operators are nested? I think it might be best to simplify things by just always returning exactly the set of input columns. I think that will "just work" - a projection will be added on top if necessary to remove extra columns after the lock. It won't be totally efficient, but that's probably ok.

If I missed something and the nested-locks case isn't a bug, the above request is just a nit that we can worry about later.
</blockquote></details>

You're correct, it wasn't working. This was one of the things I was fighting with. It was also causing problems with `ORDER BY` when the ordering column was in extra cols. Changed back to `colSetWithExtraCols`.


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (approved):
I'll take another look tomorrow morning, but this <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> pending reversion of the `MemoizeLookupJoin` change and some interesting tests.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/111662#-:-NgNsEomBdvylWmKIVKA:b-degj3w)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mgartner and @michae2)

___
*[`pkg/sql/opt/exec/execbuilder/mutation.go` line 1180 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/111662#-NgLuLrA-wlrfxxwhtlu:-NgNrOKF3LznX-OkmuNO:bi7knsh) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ec4ae291b0f316f43449610f2eb5558641d5c9d/pkg/sql/opt/exec/execbuilder/mutation.go#L1180)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Oh, nice. Done.
</blockquote></details>

Sorry, it looks like I misled you, building the props is causing NPE's because some fields are initialized. I think your initial code is ok; the group is only used to get the memo (for some reason) anyway.


<!-- Sent from Reviewable.io -->

michae2 commented:
I'm going to go for it, and we'll see where things are at when the dust settles tomorrow.

Thank you, Drew!!

bors r=DrewKimball

michae2 commented:
Once more, with feeling:

bors r=DrewKimball

michae2 commented:
bors r=DrewKimball

DrewKimball reviewed (commented):
> Reviewed 10 of 36 files at r10, 2 of 13 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/111662)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner and @michae2)

___
*[`pkg/sql/opt/optbuilder/testdata/select_for_update` line 681 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/111662#-NgPHr8XCb6KJzhz93dx:-NgPHdeKCNT4NubhUswu:b-f1pvwu) ([raw file](https://github.com/cockroachdb/cockroach/blob/97d0582c752c2d6d36cdfa66f11d745ac221c874/pkg/sql/opt/optbuilder/testdata/select_for_update#L681)):*
> ```
> 
> build set=optimizer_use_lock_op_for_serializable=true
> SELECT (SELECT a FROM t) FOR UPDATE
> ```

Why doesn't this one lock a?


<!-- Sent from Reviewable.io -->

michae2 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/111662)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @DrewKimball and @mgartner)

___
*[`pkg/sql/opt/optbuilder/testdata/select_for_update` line 681 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/111662#-NgPHr8XCb6KJzhz93dx:-NgPK_wcDbNDjBsHXuuD:bg71c0o) ([raw file](https://github.com/cockroachdb/cockroach/blob/97d0582c752c2d6d36cdfa66f11d745ac221c874/pkg/sql/opt/optbuilder/testdata/select_for_update#L681)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Why doesn't this one lock a?
</blockquote></details>

FOR UPDATE only locks tables in the FROM clause. This is a scalar subquery in the SELECT, so it is not locked by the FOR UPDATE. (This matches PostgreSQL, and also matches what we were doing before.)


<!-- Sent from Reviewable.io -->

michae2 commented:
blathers backport 23.2
